File: backend/internal/model/producto.go
package model

import (
	"time"
)

type Producto struct {
	ID            string    `json:"id"`
	Codigo        string    `json:"codigo"`
	CodigoBarras  *string   `json:"codigo_barras"`
	Nombre        string    `json:"nombre"`
	Descripcion   *string   `json:"descripcion"`
	CategoriaID   *string   `json:"categoria_id"`
	CategoriaNombre string  `json:"categoria_nombre,omitempty"`
	Precio        float64   `json:"precio"`
	StockActual   int       `json:"stock_actual"`
	StockMinimo   int       `json:"stock_minimo"`
	Imagen        *string   `json:"imagen"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

type CreateProductoRequest struct {
	Codigo        string  `json:"codigo" validate:"required"`
	CodigoBarras  *string `json:"codigo_barras"`
	Nombre        string  `json:"nombre" validate:"required"`
	Descripcion   *string `json:"descripcion"`
	CategoriaID   *string `json:"categoria_id"`
	Precio        float64 `json:"precio" validate:"required"`
	StockActual   int     `json:"stock_actual"`
	StockMinimo   int     `json:"stock_minimo"`
}

type UpdateProductoRequest struct {
	Codigo        *string `json:"codigo"`
	CodigoBarras  *string `json:"codigo_barras"`
	Nombre        *string `json:"nombre"`
	Descripcion   *string `json:"descripcion"`
	CategoriaID   *string `json:"categoria_id"`
	Precio        *float64 `json:"precio"`
	StockMinimo   *int    `json:"stock_minimo"`
}

File: frontend/stock_app/lib/domain/entities/producto.dart
class Producto {
  final String id;
  final String codigo;
  final String? codigoBarras;
  final String nombre;
  final String? descripcion;
  final String? categoriaId;
  final String? categoriaNombre;
  final double precio;
  final int stockActual;
  final int stockMinimo;
  final String? imagen;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Producto({
    required this.id,
    required this.codigo,
    this.codigoBarras,
    required this.nombre,
    this.descripcion,
    this.categoriaId,
    this.categoriaNombre,
    required this.precio,
    required this.stockActual,
    required this.stockMinimo,
    this.imagen,
    this.createdAt,
    this.updatedAt,
  });

  bool get tieneStockBajo => stockActual < stockMinimo;
  
  Producto copyWith({
    String? id,
    String? codigo,
    String? codigoBarras,
    String? nombre,
    String? descripcion,
    String? categoriaId,
    String? categoriaNombre,
    double? precio,
    int? stockActual,
    int? stockMinimo,
    String? imagen,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Producto(
      id: id ?? this.id,
      codigo: codigo ?? this.codigo,
      codigoBarras: codigoBarras ?? this.codigoBarras,
      nombre: nombre ?? this.nombre,
      descripcion: descripcion ?? this.descripcion,
      categoriaId: categoriaId ?? this.categoriaId,
      categoriaNombre: categoriaNombre ?? this.categoriaNombre,
      precio: precio ?? this.precio,
      stockActual: stockActual ?? this.stockActual,
      stockMinimo: stockMinimo ?? this.stockMinimo,
      imagen: imagen ?? this.imagen,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}